cluster-autoscaler/cloudprovider/rancher/rancher_manager.go

```diff
@@ -0,0 +1,362 @@
+	}
+
+	return clusters.Data[0].ID, nil
+}
+
+// GetCachedNodeGroups returns a list of known NodeGroup objects
+func (m *RancherManager) GetCachedNodeGroups() ([]cloudprovider.NodeGroup, error) {
+	nodeGroups := make([]cloudprovider.NodeGroup, 0)
+
+	for _, nodePool := range m.nodePoolCacheByID {
+		if hasAutoScalingEnabled(nodePool) {
+			nodeGroups = append(nodeGroups, &rancherNodeGroup{
+				manager: m,
+				id:      nodePool.ID,
+			})
+		}
+	}
+
+	return nodeGroups, nil
+}
+
+// GetCachedNodePool returns a NodePool object given its id
+func (m *RancherManager) GetCachedNodePool(id string) (*v3.NodePool, error) {
+	np, exists := m.nodePoolCacheByID[id]
+	if !exists {
+		return nil, fmt.Errorf("node pool %s not found", id)
+	}
+
+	return np, nil
+}
+
+// IncreasePoolSize increases the size of the pool identified by id by delta
+func (m *RancherManager) IncreasePoolSize(id string, delta int) error {
+	// we need a fresh node pool object instead of a cached one
+	np, err := m.nodePools.ByID(id)
+	if err != nil {
+		return err
+	}
+
+	targetQuantity := np.Quantity + int64(delta)
+	klog.Infof("Increasing %s capacity from %d to %d", np.HostnamePrefix, np.Quantity, targetQuantity)
+
+	update := map[string]interface{}{
+		v3.NodePoolFieldQuantity: targetQuantity,
+	}
+
+	_, err = m.nodePools.Update(np, update)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteNodes deletes multiple nodes from the cluster
+func (m *RancherManager) DeleteNodes(nodePoolId string, nodes []*v1.Node) error {
+	for _, node := range nodes {
+		if err := m.DeleteNode(nodePoolId, node); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeleteNode deletes a single node from the cluster
+func (m *RancherManager) DeleteNode(nodePoolId string, node *v1.Node) error {
+	rancherNode, err := m.GetNodeForKubernetesNode(node.Name)
+	if err != nil {
+		return err
+	}
+
+	if rancherNode == nil || rancherNode.NodePoolID != nodePoolId {
+		return fmt.Errorf("node %s not found", node.Name)
+	}
+
+	klog.Infof("Marking node %s (%s) for deletion", rancherNode.ID, rancherNode.NodeName)
+	err = m.nodes.ActionPleaseKillMe(rancherNode)
+	if err != nil {
+		return err
+	}
+
+	// cluster autoscaler specifies that we should wait until the node disappears from the underlying infrastructure
+	// before proceeding so we'll park here for a while
+	for retry := 0; retry < nodeDeletionVerifierRetryLimit; retry++ {
+		time.Sleep(nodeDeletionVerifierRetryDelay)
+
+		_, err := m.nodes.ByID(rancherNode.ID)
+		if err == nil {
+			continue
+		}
+
+		if clientbase.IsNotFound(err) {
+			// yay :)
+			return nil
+		}
+
+		klog.Warningf("unexpected error while checking node deletion (attempt %d/%d): %s",
+			retry, nodeDeletionVerifierRetryLimit, err)
+
+	}
+
+	return fmt.Errorf("unable to confirm node %s (%s) deletion for %s",
+		rancherNode.Name, rancherNode.ID, nodeDeletionVerifierRetryLimit*nodeDeletionVerifierRetryDelay)
```

Comment:
[nit] when you announced the deletion in a log before this one, you used this order of id and name
```suggestion
		rancherNode.ID, rancherNode.Name, nodeDeletionVerifierRetryLimit*nodeDeletionVerifierRetryDelay)
```